Ilcm (#49)
GovTechSG/stackops-docs: ilcm.md

| **Primary Shard** 	| Contains either the entire or a subset of data per index, based on configuration.	|
| **Replica Shard** 	| Replicates the respective primary shard. It takes over the role if the primary shard becomes unavailable. |

?> For example, if you have 1 GB of CloudWatch index and configured it with 3 primary and replica shards, each primary and their respective replica shard will have 333 MB of CloudWatch index.

### Additional resources

- [Stack Overflow](https://stackoverflow.com/questions/15694724/shards-and-replicas-in-elasticsearch)
- [Coding Explained](https://codingexplained.com/coding/elasticsearch/understanding-sharding-in-elasticsearch)
- [Opster Blogs](https://opster.com/blogs/elasticsearch-shards-and-replicas-getting-started-guide/)
- [Elastic Documentation](https://www.elastic.co/guide/en/elasticsearch/reference/8.8/size-your-shards.html#shard-size-recommendation)

## What does Index Lifecycle Policy do?

Index Lifecycle Policy(ILP) determines the shard movement within data nodes. Since ILP is highly reusable, it can be applied to an individual index or a data stream of indices.

> created once and applied to all indices in a data-stream, making it reusable across multiple data-streams. 

Author: done.

# Index Lifecycle Management

Index Lifecycle Management automates the lifecycle of Elasticsearch indexes, which includes their creation, management, and deletion.

## What is Index Lifecycle Management?

To ensure smooth data management and avoid storage issues, you need to monitor your data ingestion rate as it continues to grow over time. Uncontrolled ingestion without monitoring can quickly fill up your storage capacity, leading to unexpected shortages and potential data loss or system disruptions. Failure to address this issue may result in the loss of access to Kibana. For more information, refer to [Watermark limits](/getting-started/watermark-limit).

> To effectively manage and assess deployment health, you need to configure your [Index Lifecycle Policies](#index-lifecycle-policy) as required. Index Lifecycle Management (ILM) allows you to understand your current and future ingestion requirements, helping you to make informed decisions on deployment sizing.

Author: sorry this is the outdated file, but I have used "improve"  in the following paragraph.

Hence, tenants who have set up their Elastic Agent integrations for logs and metrics ingestion need to review their Index Lifecycle Policies. Index Lifecycle Management (ILM) enables tenants to effectively manage and **improve** the health of their deployments. With proper ILM strategy, tenants can understand their current and future ingestion requirements, and help them make informed decisions regarding their deployment sizing.

## Overview of indices and shards

The following diagram gives a high-level definition of **Index**, **Primary Shard** and **Replica Shard**.

![shards and indices](images/ilm-shards.png)

| Terms 	| Definition 	|
|---	|---	|
| **Integration** 	| Defines how an index is created and how data is ingested by the Elastic Agent. In this scenario, CloudWatch and Kubernetes are used for the integration, resulting in 2 indices. 	|
| **Index** 	| This is how Elasticsearch logically groups the dataset and stores them as shards. Each index can have multiple shards. 	|
| **Shards** 	| Subset of data within an Elasticsearch index. There are two types : Primary and Replica shards. 	|
| **Primary Shard** 	| Contains either the entire or a subset of data per index, based on configuration.	|
| **Replica Shard** 	| Replicates the respective primary shard. It takes over the role if the primary shard becomes unavailable. |

> ?> For example, if you have 1 GB of CloudWatch index and configured it with 3 primary and replica shards, each primary and replica shard will have 333 MB of CloudWatch index.

Author: done

### Additional resources

- [Stack Overflow](https://stackoverflow.com/questions/15694724/shards-and-replicas-in-elasticsearch)
- [Coding Explained](https://codingexplained.com/coding/elasticsearch/understanding-sharding-in-elasticsearch)
- [Opster Blogs](https://opster.com/blogs/elasticsearch-shards-and-replicas-getting-started-guide/)
- [Elastic Documentation](https://www.elastic.co/guide/en/elasticsearch/reference/8.8/size-your-shards.html#shard-size-recommendation)

## Index Lifecycle Policy

> Index Lifecycle Policy(ILP) determines shard movement within data nodes. It is created once and applied to all indices in a data-stream, making it reusable across multiple data-streams. 

Reviewer: ```suggestion
Index Lifecycle Policy (ILP) determines shard movement within data nodes. It is highly reusable and can be applied to any individual index or a data stream of indices```

ILPs can consist the following five phases that define the conditions for data stream movement between nodes.

1. [Hot phase](#hot-phase)
2. [Warm phase](#warm-phase)
3. [Cold phase](#cold-phase)
4. [Frozen phase](#frozen-phase)
5. [Delete phase](#delete-phase)

The following diagram depicts the 5 phases.

![ilm-phases](images/ilm-phases.png)

## Configure Index Lifecycle Policy

> StackOps provides the default ILP templates: **logs** and **metrics** for a deployment. You can modify the configurations for your deployment as needed. 

Reviewer: ```suggestion
StackOps provides the default ILPs: **logs** and **metrics** for a deployment. You can modify the configurations for your deployment as needed. 
```

Author: done.

?>**What values to specify in the Rollover configuration**?<br>Assuming you are using two integrations, each creating an index on their own which are indices A and B.<br><br>- For the **Maximum primary shard size**, use the average daily data ingestion. If index A and B ingest 10 GB and 30 GB per day respectively, the average is 20 GB. Therefore, set it to 20 GB.<br><br>- If you want to keep your indices on the hot node for 7 days before shifting to the **Warm phase**, set the **Maximum age** as 7 days.<br><br>**How do these configurations work?**<br>These configurations work like an **OR** condition. If the data in index-a-001 reaches 20 GB before 7 days, a new index (index-a-002) is created, and subsequent data is directed to index-a-002. If index-a-001 was created on 1 January 2023 and the data ingestion has not reached 20 GB by 8 January 2023, a new index (index-a-002) is created, and subsequent data is directed there. The old index(index-a-001) stays in the **Hot phase** for the specified **Maximum age**, which is 7 days in this example and then it moves to the Warm phase. 

3. Click **Learn more** to view the additional configuration details.

### Warm phase

> 

Reviewer: Put disclaimer here

----- If Rollover is Enabled -----
Please be warned that when Rollover settings are configured in the [Hot Phase](https://confluence.ship.gov.sg/pages/viewpage.action?spaceKey=STO&title=Index+Lifecycle+Management#IndexLifecycleManagement-ilm-ho), we recommend that the Rollover Age should not be longer than the Warm Phase Age. This will result in data staying longer than expected in the Hot Node as the Phase Shift between Hot → Warm will only happen once the [Rollover has been completed](https://www.elastic.co/guide/en/elasticsearch/reference/current/ilm-index-lifecycle.html#ilm-phase-transitions).

Author: done

   
   ?> **What value to specify in the warm phase**?<br>We recommend you to update this value based on the following factors:<br>- Size of your hot tier data nodes.<br>- Your daily data ingestion rate. Refer to the following use case to understand this better.<br><br>Assume your hot node has a capacity to store 60 GB of data and your daily data ingestion rate is 5 GB with a [**watermark limit**](watermark-limit) 85%-90%. This means your hot nodes can store approximately 54 GB (90% of 60 GB) of data before triggering a warning. In other words, it means you can store up to 10 days worth of data. However, it's important to consider potential spikes in data flow on any given day. To account for such scenarios, it's recommended to include a buffer period. For this use case, we suggest setting the **Move data into phase when** configuration to 7 days.

2. Click **Learn more** to view the additional configuration details.

### Cold phase

Similar to the **Warm phase**, in the **Cold phase** configuration, we focus on the settings of **Move data into phase when** field. This defines when the indices move to the **Cold phase**.

> - In this phase, indices are stored as [Snapshots](https://www.elastic.co/guide/en/elasticsearch/reference/current/searchable-snapshots.html), resulting in lower cost and performance.

Reviewer: ```suggestion
- In this phase, optionally, indices can be stored as [Snapshots](https://www.elastic.co/guide/en/elasticsearch/reference/current/searchable-snapshots.html), resulting in lower cost and performance.
```

2. Click **Learn more** to view the additional configuration details.

### Delete phase

As the Indices grow larger by the seconds, storage capacity issues are inevitable. **Delete phase** is the last phase where indices are deleted after a certain duration. Indices enter this phase based on the settings for the **Move data into phase when** field. 

?> To know more about other settings, click **Learn more** or you may also read about [Elastic Shard Sizing ](https://www.elastic.co/guide/en/elasticsearch/reference/8.8/size-your-shards.html#shard-size-recommendation).

**To edit the delete phase configurations**

> 1. In the **Delete phase** configurations section, specify the required duration in the **Move data into phase when** field. After this duration, the indices move from **frozen** node to **delete** node.

Reviewer: ```suggestion
1. In the **Delete phase** configurations section, specify the required duration in the **Move data into phase when** field. After this duration, the indices will be **deleted**.
```

   ![ilm-delete](images/ilm-delete.png)

>    As a recommendation, StackOps has preconfigured this duration to 365 Days. Indices that are 365 days old move to the delete node.  

Reviewer: ```suggestion
   As a recommendation, StackOps has preconfigured this duration to 365 Days. Indices older than 365 days old are deleted and will not be recoverable.  
```

### FAQs

<details>
<summary>How long should I retain my indices?</summary>
  <!-- leave an empty line *️⃣  -->
  

> Indices reside on different phases depending on your index lifecycle policy configurations. So, while defining the indices duration, consider the time when the data ingestion starts at the Hot phase until the data transitions to the Delete Phase.

Reviewer: ```suggestion
Indices reside on different phases depending on your index lifecycle policy configurations. So, while defining the indices duration, consider the time when the data ingestion starts at the Hot phase until it ends in the Delete Phase.
```

Author: Done

 <!--
 While defining the indices duration, consider the time when the data ingestion starts at the Hot phase until the data transitions to the Delete Phase.-->

</details>
<!-- leave an empty line *️⃣  -->

<details>
<summary>How long should the data reside on the indices?</summary>
  <!-- leave an empty line *️⃣  -->

This can be estimated by considering the following:

- Daily date ingestion rate
- Size of your nodes on the StackOps portal

Based on these information, while configuring your Index Lifecycle Policy, specify the required duration for which you want to retain the data on the indices in each phase.

</details>

<details>
<summary>How much storage is required for the indices to be readily available with reasonable search time?</summary>
  <!-- leave an empty line *️⃣  -->
 

It is important to weigh your options on how much storage is required for the indices in each phase.

![index search performance vs cost](images/ilm-perf.png)

</details>

<details>

> <summary>What should be the size of my warm nodes?</summary>

Reviewer: ```suggestion
<summary>What should be the size of my nodes?</summary>
```

  <!-- leave an empty line *️⃣  -->

We recommend you to make this decision by considering the following:

- Your daily date ingestion rate.
- The duration for which your indices need to remain on the warm nodes. Refer to the following use case for a better understanding.

> Let's assume that your indices transition from the **Hot phase** to the **Warm phase** after 7 days and move to the **Cold phase** after 30 days. This implies that the indices stay in the **Warm phase** for 23 days (30 days - 7 days = 23 days). If your daily data ingestion rate is 5 GB, it is recommended to allocate at least 115 GB of storage for your warm nodes (5 GB * 23 days = 115 GB).

Reviewer: ```suggestion
Take warm node as an example. Let's assume that your indices transition from the **Hot phase** to the **Warm phase** after 7 days and move to the **Cold phase** after 30 days. This implies that the indices stay in the **Warm phase** for 23 days (30 days - 7 days = 23 days). If your daily data ingestion rate is 5 GB, it is recommended to allocate at least 115 GB of storage for your warm nodes (5 GB * 23 days = 115 GB).

Continuing the example with frozen node. Let's assume that your indices transition from the **Cold phase** to the **Frozen phase** after 90 days and move to the **Delete phase** after 365 days. This implies that the indices stay in the **Frozen phase** for 275 days (365 days - 90 days = 275 days). If your daily data ingestion rate is 5 GB, it is recommended to allocate at least 1375 GB of storage for your frozen nodes (5 GB * 275 days = 1375 GB).
```

Author: done

</details>
<!-- leave an empty line *️⃣  -->

<details>

> <summary>What should be the size of my frozen nodes?</summary>

Reviewer: ```suggestion
```
Remove this last section and combine it with the previous example so users can make the connection

Author: done

  <!-- leave an empty line *️⃣  -->

We recommend you to make this decision by considering the following:

- Your daily data ingestion rate.
- The duration for which your indices need to remain on the frozen nodes. Refer to the following use case for a better understanding.

> Let's assume that your indices transition from the **Hot phase** to the **Warm phase** after 7 days. When your indices are 30 days old, they move to the **Cold phase** and when they are 90 days old, they move to the **Frozen phase**. When the indices are 365 days old, they move from **Frozen phase** to **Delete phase**. This implies that the indices stay in the **Frozen phase** for 60 days (365 days - 90 days = 275 days). If your daily data ingestion rate is 5 GB, it is recommended to allocate at least 1375 GB of storage for your frozen nodes (5 GB * 275 days = 1375 GB).

Reviewer: ```suggestion
```

</details>

<details>
<summary>What should be the size of my nodes?</summary>
  <!-- leave an empty line *️⃣  -->

We recommend you to make this decision by considering the following:

- Your daily date ingestion rate.

> - The duration for which your indices need to remain on the warm nodes. 

Author: @wathian I have removed the word "warm" but shall I tweak the text "Refer to the following scenario for a better understanding." Otherwise, there is no connectivity between the bullet points and the use case

> - The duration for which your indices need to remain on the warm nodes. 

Author: done!
